src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -642,6 +662,61 @@ ExecuteSingleSelectTask(CitusScanState *scanState, Task *task)
 
 
 /*
+ * BuildPlacementSelectList builds a list of SELECT placement accesses
+ * which can be used to call StartPlacementListConnection or
+ * GetPlacementListConnection.
+ */
+List *
+BuildPlacementSelectList(char *nodeName, int nodePort, List *relationShardList)
+{
+	ListCell *relationShardCell = NULL;
+	List *placementAccessList = NIL;
+
+	foreach(relationShardCell, relationShardList)
+	{
+		RelationShard *relationShard = (RelationShard *) lfirst(relationShardCell);
+		List *shardPlacementList = NIL;
+		ShardPlacement *placement = NULL;
+		ShardPlacementAccess *placementAccess = NULL;
+		bool missingOK = false;
+
+		shardPlacementList = FinalizedShardPlacementList(relationShard->shardId);
+
+		placement = SearchShardPlacementInList(shardPlacementList, nodeName, nodePort,
+											   missingOK);
+		if (placement == NULL)
```

Comment:
Getting the final list, which will first build the full list, and then searching in it isn't exactly elegant & causes a good chunk of allocations...